lib/openscap_parser/value.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+module OpenscapParser
+  class Value < XmlNode
+    include OpenscapParser::Util
+
+    def id
+      @id ||= parsed_xml['id']
+    end
+
+    def description
+      @description ||= newline_to_whitespace(parsed_xml.at_css('description')&.text)
+    end
+
+    def title
+      @title ||= parsed_xml.at_css('title')&.text
+    end
+
+    def type
+      @type ||= parsed_xml['type'] || 'string'
+    end
+
+    def generic_selector(type, selector = nil)
+      unless instance_variable_get("@#{type}")
+        element_name = type.to_s.sub('_', '-')
+        elements = parsed_xml.xpath(element_name).each_with_object({}) do |element, elements|
+          elements[element.at_xpath('@selector')&.text.presence] = element&.text
+        end
+        instance_variable_set("@#{type}", elements)
+      end
+
+      cache = instance_variable_get("@#{type}")
+
+      return cache[selector] if selector
+      cache[nil] || cache.values.first
+    end
```

Comment:
If you set `cache` before the `unless`, you can still overwrite it when building the `elements` hash. This also helps you preventing the [shadowing](https://stackoverflow.com/questions/6259314/what-does-shadowing-mean-in-ruby) of the `elements` variable. Ideally you should always avoid it when possible, and it's only unavoidable when you run out from variable names.

Then of course you have to set the instance variable from `cache` and not `elements` and you can get rid of the second `instance_variable_get`.

Reply:
Ahh right, good idea